### PR TITLE
Revamped AGD calculation: more efficient but more complex.

### PR DESCRIPTION
**Pull request by tmontes**
Worked on this for a bit, focusing on addressing #49.

Benefits:
* Works.
* Computationally more efficient and mostly O(1) instead of certainly O(n).

Trade-off:
* More lines of code and comments (I had expected less, but estimated wrong!).
* Code and algorithm is much harder to understand and reason about.

I'm submitting this PR for the sake of sharing and completeness. But I recommend against merging it, given that for the purposes, the current, simpler, implementation is perfectly adequate.

In other words:
* Take a look at it if you want.
* Reject this PR and cancel #49.

What do you say?

### CODE REVIEW

**Comment by nununo** [comment]
Rejected because it would add unnecessary complexity with little performance gains
